File: scripts/mazes.py
'''
Working through different maze algorithms from the book "Mazes for Programmers" by Jamis Buck
https://www.amazon.com/Mazes-Programmers-Twisty-Little-Passages/dp/1680500554?ref_=ast_sto_dp

Outputs to terminal or writes a PNG file representing the maze
'''

from PIL import Image, ImageDraw
import random

class Distances:

    def __init__(self, root):
        self.root = root
        self.cells = {}
        # Distance from root to itself is 0
        self.cells[self.root] = 0

    def __index__(self, cell):
        return self.cells[cell]
    
    def __index__(self, cell, distance):
        self.cells[cell] = distance

    def get_cells(self):
        return self.cells.keys()

class Cell:
    
    def __init__(self, row, column):
        self.north = None
        self.south = None
        self.east = None
        self.west = None
        self.row = row
        self.column = column
        self._links = {}
        
    def link(self, cell, bidi=True):
        self._links[cell] = True
        if bidi == True:
            cell.link(self, False)
    
    def unlink(self, cell, bidi=True):
        del self._links[cell]
        if bidi == True:
            cell.unlink(self, False)
    
    def links(self):
        self._links
        
    def is_linked(self, cell):
        if cell in self._links:
            return True
        else:
            return False
    
    def get_neighbors(self):
        neighbors = []
        if self.north:
            neighbors.append(self.north)
        if self.south:
            neighbors.append(self.south)
        if self.east:
            neighbors.append(self.east)
        if self.west:
            neighbors.append(self.west)
    
    def distances(self):
        d = Distances(self)
        frontier = [self]
        while frontier:
            new_frontier = []
            for cell in frontier:
                for linked in cell.links():
                    if d[linked]:
                        d[linked] = d[cell] + 1
                        new_frontier.append(linked)
            frontier = new_frontier
        return d
            
class Grid:
    
    def __init__(self, rows, columns):
        self._rows = rows
        self._columns = columns
        self.prepare_grid()
        
    def prepare_grid(self):
        self._grid = []
        for row in range(self._rows):
            self._grid.append([])
            for col in range(self._columns):
                self._grid[row].append(Cell(row, col))
        self.configure_cells()
        
    # Configure 2-dimensional grid starting with the top-leftmost cell,
    #   and count left-right and then downward (similar to an SQL table!)
    def configure_cells(self):
        for cell in self.each_cell():
            row, col = cell.row, cell.column
            cell.north = self.at(row - 1, col)
            cell.south = self.at(row + 1, col)
            cell.west = self.at(row, col - 1)
            cell.east = self.at(row, col + 1)

    def at(self, row, column):
        if row < 0 or row >= self._rows:
            return None
        elif column < 0 or column >= len(self._grid[row]):
            return None
        return self._grid[row][column]
    
    def random_cell(self):
        random_row = random.randint(0, self._rows)
        random_col = random.randint(0, len(self._grid[random_row - 1]))
        return random.choice(self._grid)
    
    def size(self):
        return self._rows * self._columns
    
    def each_row(self):
        for row in range(len(self._grid)):
            yield self._grid[row]
    
    def each_cell(self):
        for grid_row in self.each_row():
            for cell in grid_row:
                if cell != None:
                    yield cell
                
    def __str__(self):
        output = "+" + "---+" * self._columns + "\n"
        for row in self.each_row():
            # Top and Bottom of Cell Wall
            top = "|"
            bottom = "+"
            for cell in row:
                if cell is None:
                    cell = Cell(-1, -1)
            
                body = "   "
                if cell.is_linked(cell.east) is True:
                    east_boundary = " "
                else:
                    east_boundary = "|"
                top += body + east_boundary
                
                if cell.is_linked(cell.south) is True:
                    south_boundary = "   "
                else:
                    south_boundary = "---"
           
                corner = "+"
                bottom += south_boundary + corner

            output += top + "\n"
            output += bottom + "\n"
        return output

class BinaryTree:
    
    # Count left-right, bottom-top
    @staticmethod
    def on(grid):
        for cell in grid.each_cell():
            neighbors = []
            if cell.north is not None:
                neighbors.append(cell.north)
            if cell.east is not None:
                neighbors.append(cell.east)
            
            if len(neighbors) != 0:
                random_index = random.randrange(len(neighbors))
                neighbor = neighbors[random_index]
                if neighbor is not None:
                    cell.link(neighbor)

# class DFS:
#     @staticmethod
#     def on(grid):
#         for cell in grid.each_cell():
#             maze_stack = [cell]
#             while maze_stack:
#                 current = maze_stack.pop()
#                 # Find all unvisited neighbors of current cell
#                 neighbors = [neighbor for neighbor in current if not neighbor.links()]
#                 if neighbors:
#                     # Push unvisited cells with respcet to current
#                     maze_stack.append(current)
#                     neighbor = random.choice(neighbors)
#                     current.link(neighbor)
#                     maze_stack.append(neighbor)

'''
Utility function to draw the maze using PIL
'''
def draw_maze(grid, cell_size=10):
    img_width = grid._columns * cell_size
    img_height = grid._rows * cell_size
    background = (255, 255, 255)
    wall = (0, 0, 0)

    img = Image.new("RGB", (img_width + 1, img_height + 1), background)
    draw = ImageDraw.Draw(img)

    for cell in grid.each_cell():
        x1 = cell.column * cell_size
        y1 = cell.row * cell_size
        x2 = (cell.column + 1) * cell_size
        y2 = (cell.row + 1) * cell_size

        if cell.north is None:
            draw.line((x1, y1, x2, y1), fill=wall)
        if cell.west is None:
            draw.line((x1, y1, x1, y2), fill=wall)
        if not cell.is_linked(cell.east):
            draw.line((x2, y1, x2, y2), fill=wall)
        if not cell.is_linked(cell.south):
            draw.line((x1, y2, x2, y2), fill=wall)

    return img

if __name__ == '__main__':
    
    CELL_SIZE = 4
    grid = Grid(10, 10)
    DFS.on(grid)
    
    # print(str(grid))
    img = draw_maze(grid, cell_size=CELL_SIZE)
    img.save("dfs.png")
    

File: scripts/mazes.rb
# Implementing maze generation algorithms in Ruby

require 'chunky_png'

# Measure distances between grid cells
class Distances

    def initialize(root)
        @root = root
        @cells = {}
        @cells[@root] = 0
    end

    def [](cell)
        @cells[cell]
    end

    def []=(cell, distance)
        @cells[cell] = distance
    end

    def cells
        @cells.keys
    end

    def path_to(goal)
        current = goal
        breadcrumbs = Distances.new(@root)

        breadcrumbs[current] = @cells[current]

        until current == @root
            current.links.each do |neighbor|
                if @cells[neighbor] < @cells[current]
                    breadcrumbs[neighbor] = @cells[neighbor]
                    current = neighbor
                    break
                end
            end
        end
        breadcrumbs
    end

    def max
        max_distance = 0
        max_cell = @root
        @cells.each do |cell, distance|
            if distance > max_distance
                max_cell = cell
                max_distance = distance
            end
        end
        [max_cell, max_distance]
    end
end

class Cell
    attr_reader :row, :column
    attr_accessor :north, :south, :east, :west

    def initialize(row, column)
        @row, @column = row, column
        @links = {}
    end

    # Helper methods for linking and unlinking cells in mazes
    def link(cell, bidi=true)
        @links[cell] = true
        cell.link(self, false) if bidi
        self
    end

    def unlink(cell, bidi=true)
        @links.delete(cell)
        cell.unlink(self, false) if bidi
        self
    end

    def links
        @links.keys
    end

    def linked?(cell)
        @links.key?(cell)
    end

    def neighbors
        list = []
        list << north if north
        list << south if south
        list << east if east
        list << west if west
        list
    end

    def to_s
        "row: #{@row} , column: #{@column}"
    end

    # Compute distances between cell starting points
    # using Dijkstra's algo
    def distances
        distances = Distances.new(self)
        frontier = [self]
        while frontier.any?
            new_frontier = []
            frontier.each do |cell|
                cell.links.each do |linked|
                    next if distances[linked]
                    distances[linked] = distances[cell] + 1
                    new_frontier << linked
                end
            end
            frontier = new_frontier
        end
        distances
    end
end

class Grid
    attr_reader :rows, :columns

    def initialize(rows, columns)
        @rows = rows
        @columns = columns
        @grid = prepare_grid

        configure_cells

    end

    def prepare_grid
        Array.new(rows) do |row|
            Array.new(columns) do |column|
                Cell.new(row, column)
            end
        end
    end

    # Configure 2-dimensional grid starting with the top-leftmost cell,
    #   and count left-right and then downward (similar to an SQL table!)
    def configure_cells
        each_cell do |cell|
            row, col = cell.row, cell.column
            cell.north = self[row - 1, col]
            cell.south = self[row + 1, col]
            cell.west = self[row, col - 1]
            cell.east = self[row, col + 1]
        end
    end

    def [](row, column)
        return nil unless row.between?(0, @rows - 1)
        return nil unless column.between?(0, @grid[row].count - 1)
        @grid[row][column]
    end

    def random_cell
        row = rand(@rows)
        column = rand(@grid[row].count)
        self[row, column]
    end

    def size
        @rows * @columns
    end
    
    def each_row
        @grid.each do |row|
            yield row
        end
    end

    def each_cell
        each_row do |row|
            row.each do |cell|
                yield cell if cell
            end
        end
    end

    def contents_of(cell)
        " "
    end

    def background_color_for(cell)
        nil
    end

    # ASCII-character display of maze
    def to_s
        output = "+" + "---+" * columns + "\n"
        each_row do |row|
            # Top and Bottom of Cell Wall
            top = "|"
            bottom = "+"
            row.each do |cell|
                cell = Cell.new(-1, -1) unless cell
                # body = "   "
                body = " #{contents_of(cell)} "
                east_boundary = (cell.linked?(cell.east) ? " " : "|")
                top << body << east_boundary
                south_boundary = (cell.linked?(cell.south) ? "   " : "---")
                corner = "+"
                bottom << south_boundary << corner
            end
            output << top << "\n"
            output << bottom << "\n"
        end
        output
    end

    def to_png(cell_size: 25)
        img_width = cell_size * columns
        img_height = cell_size * rows

        background = ChunkyPNG::Color::WHITE

        wall = ChunkyPNG::Color::BLACK

        img = ChunkyPNG::Image.new(img_width + 1, img_height + 1, background)

        [:backgrounds, :walls].each do |mode|
            each_cell do |cell|
                x1 = cell.column * cell_size
                y1 = cell.row * cell_size
                x2 = (cell.column + 1) * cell_size
                y2 = (cell.row + 1) * cell_size

                if mode == :backgrounds
                    color = background_color_for(cell)
                    img.rect(x1, y1, x2, y2, color, color) if color
                else
                    img.line(x1, y1, x2, y1, wall) unless cell.north
                    img.line(x1, y1, x1, y2, wall) unless cell.west
                
                    img.line(x2, y1, x2, y2, wall) unless cell.linked?(cell.east)
                    img.line(x1, y2, x2, y2, wall) unless cell.linked?(cell.south)
                end
            end
        end
        img
    end
end

class DistanceGrid < Grid
    attr_accessor :distances

    def contents_of(cell)
        if distances && distances[cell]
            distances[cell].to_s(36)
        else
            super
        end
    end
end

class ColoredGrid < Grid
    def distances=(distances)
        @distances = distances
        farthest, @maximum = distances.max
    end

    def background_color_for(cell)
        if !@distances[cell]
            return nil
        end
        distance = @distances[cell]
        # or return nil?
        intensity = (@maximum - distance).to_f / @maximum
        dark = (255 * intensity).round
        bright = 128 + (127 * intensity).round
        ChunkyPNG::Color::rgb(dark, bright, dark)
    end
end

class BinaryTree
    def self.on(grid)
        grid.each_cell do |cell|
            neighbors = []
            neighbors << cell.north if cell.north
            neighbors << cell.east if cell.east
            index = rand(neighbors.length).floor
            neighbor = neighbors[index]
            cell.link(neighbor) if neighbor
        end
    end
end

class Sidewinder

    def self.on(grid)
        grid.each_row do |row|
            run = []
            row.each do |cell|
                run << cell
                at_eastern_boundary = (cell.east == nil)
                at_northern_boundary = (cell.north == nil)
                should_close_out = at_eastern_boundary || (!at_northern_boundary && rand(2) == 0)

                if should_close_out
                    member = run.sample
                    member.link(member.north) if member.north
                    run.clear
                else
                    cell.link(cell.east)
                end
            end
        end
        grid
    end

end

class AldousBroder

    def self.on(grid)
        cell = grid.random_cell
        unvisited = grid.size - 1
        while unvisited > 0
            neighbor = cell.neighbors.sample
            if neighbor.links.empty?
                cell.link(neighbor)
                unvisited -= 1
            end
            cell = neighbor
        end
        grid
    end

end

class Wilsons

    def self.on(grid)
        unvisited = []
        grid.each_cell { |cell| unvisited << cell }
        first = unvisited.sample
        unvisited.delete(first)
        while unvisited.any?
            cell = unvisited.sample
            path = [cell]
            while unvisited.include?(cell)
                cell = cell.neighbors.sample
                position = path.index(cell)
                if position
                    path = path[0..position]
                else
                    path << cell
                end
            end
            0.upto(path.length - 2) do |index|
                path[index].link(path[index + 1])
                unvisited.delete(path[index])
            end
        end
        grid
    end
end

6.times do |n|

    grid = ColoredGrid.new(35, 25)
    AldousBroder.on(grid)

    start = grid[0, 0]
    end1 = grid[grid.rows / 2, grid.columns / 2]

    grid.distances = start.distances.path_to(end1)

    filename = "ab%02d.png" % n
    grid.to_png.save(filename)
    puts "saved to #{filename}"
end